### ISSUE TITLE
Proposal: have Jekyll:Convertible raise "conversion" exceptions

### ISSUE THREAD

**Issue by e2** [jekyll/jekyll#3289]
Since "convertible" is plugin agnostic, it should raise plugin-agnostic errors.

### Current behavior (example)

`Haml` on syntax errors throws `Haml::SyntaxError`, which is a subclass of StandardError.

Jekyll catches Exception based exceptions to show an error - and re-raises the same exception:

https://github.com/jekyll/jekyll/blob/43a28aed967b66de87e2a539ca92c5cb93ce9b66/lib/jekyll/convertible.rb#L71

### The Problem 

While this is useful for people (catching exception to show the message and re-raise it), it is useless for Jekyll API consumers, because they have no way of knowing if the error was a result of a plugin conversion gone wrong (and not some bad system exception).

### Expected Behavior

Jekyll should wrap non-generic exceptions into a `ConversionError` exception, so apps using Jekyll API can ignore a range of related errors ("conversion errors") - without having to "catch everything" and "guess if it's just a syntax error a plugin reports".

E.g. in the example above, Jekyll should throw a `Jekyll::ConversionError` with a method `orignal_exception` set to the caught `Haml::SyntaxError` exception instance.

### The idea

If Convertible is a wrapper for plugins, it should also wrap exceptions it forwards.

**Comment by parkr** [comment]
This is a Good Idea :tm:. We're trying to move everything to a Renderer-powered Document instead of the Convertible class, but yes the abstraction should still stand.

**Comment by envygeeks** [comment]
IMO this is a broader issue if I am reading into this right.  Sure we don't raise conversion errors, but what happens with the hundreds of other errors that can happen? This does not solve those, what we need is a generic JekyllError wrapper that the API spits out and the command interface needs to consume and decide what to do with that based on 1. your options 2. a flag on the command system.  This would also simplify and solve another issue that has a similar problem involving TravisCI.  Our API's should not make the assumption that they are being used by our command system and are obviously way too coupled to that assumption.

**Comment by e2** [comment]
TL;DR - Jekyll needs to implement 3 kinds of error handling: build errors, conversion errors and plugin-not-wrapping-exceptions errors. This would allow any lib to reasonably use any level of the Jekyll API.

@envygeeks 

> but what happens with the hundreds of other errors that can happen?

Well:
1. Jekyll *knows* it is a conversion error
2. Jekyll "displays" the error as a "Conversion Error" (so it's technically handling it and calling it what it is)
3. Now the client library using Jekyll has to catch pretty much `Exception` and guess (by parsing the message I guess) whether the error is something that is perfectly fine (Haml::Syntax error) or if it should terminate immediately (not enough disk space when Jekyll tried to write the output).

So instead of using a tree of exceptions (Exception, StandardError, RuntimeError, SomeCustomError, etc.), Jekyll is forcing client libraries to work out the exception type based on either the string of the exception, or catching *all* possible exception types in all loaded gems and jekyll plugins.

Simply put, the information that an exception happened during "normal" Jekyll conversion is lost. If jekyll fails during a build, it should result in a build error, not a "something went wrong during something being done" exception that only a human being can understand.

The client library doesn't have the choice of handling only conversion/build errors.

Of course, the ideal situation would be for jekyll plugins to catch the exceptions and throw them as a specific Jekyll type (since they subclass Jekyll::Converter anyway).

E.g. here: https://github.com/samvincent/jekyll-haml/blob/master/lib/jekyll-haml.rb#L21

In this example, a Haml::Syntax error should at least throw a Jekyl::ConverterError (or something) instead of the Haml::Syntax error.

So, 3 things should be handled:
1. Any build error - wrapped and handled by Jekyll::Build error or something
2. Any conversion error - wrapped and handled by Jekyll (e.g. Jekyll::UnknownConverterError or LiquidError, etc.)
3. Errors handled by plugins (should be either wrapped in Jekyll::Error or reported as bugs in the plugin with a message about properly handling and raising Jekyll exceptions)

**Comment by envygeeks** [comment]
I think a single wrapper Jekyll::Error with a flag (`#type`) should be sufficient for you to track where and how with minimal encapsulation with maximum information and wrapability.  Using a single wrapper for everything we wrap around allows you to create a single wrapper yourself for dozens of problems and then decide based on the flags where they should go and what you should do.  Maybe even provide a `#level` flag that lets you know what we rate it at.

**Comment by envygeeks** [comment]
I'm not against multiple error wrappers though, even if I do think these kinds of errors can be encapsulated cleanly in a single wrapper, I'll leave that logistical decision up to you and @parkr and @mattr- if he wants to chime in, I'm a minimalist they might not be.

**Comment by e2** [comment]
Some libs may want to use Jekyll to build, others may want just the plugin/conversion. Everything can (should) subclass or be wrapped with `Jekyll::Error` (unless it *really* is a non-jekyll-expected scenario).

I don't believe levels make sense, since that's context dependent. Even the example: Haml::Syntax error is a StandardError - which is expected. But a StandardError coming from Jekyll says something REALLY unexpected happened (e.g. NoMethodError in Jekyll's code, no _config.yml file, etc.).

For `Guard::Jekyll`, *every* Jekyll "website-development-typical" error should be caught and handled - to prevent blowing up Guard completely, leaving dead servers, and used TCP ports, child processes, etc.

Overall a '#type' method suggests the lack of object oriented programming instead of using something more idiomatic to Ruby.

'#level' as I've mentioned is context dependent - there are too many contexts here for a "global" idea of levels to make sense.

**Comment by envygeeks** [comment]
On the "lack of object orientated programming" that's a bit hyperbolic, extremely if we dig into what an object really is in Ruby: everything; and when we dig into what OOP actually means, and a type data attribute definitely falls into that definition.  Having a thousand different error classes or a single method (that is an accessor for that data attribute) that establishes the type does not define whether or not your are being OOP, but that's an argument that is not even relevant to this and will just throw this solution off into a flame war of sorts.

The `Guard::Jekyll` problem you describe is really broad. Can you elaborate? With all that said, I digress on this situation because I can see a fitting place for both solutions which means that others should and probably will come up with the final solution that is best for Jekyll and mostly because I don't think we should assume control over any error past wrapping it and choosing whether to raise it, and there is no need for for a bunch of objects to do that.

**Comment by e2** [comment]
TL;DR - "a bunch of classes" is actually proper structured exception handling - necessary for complex projects like Jekyll to be useful in many ways

> a type data attribute definitely falls into that definition

A type is not the same as an object. Also this goes against duck-typing (checking if an object responds to a method, while a type is Ruby is really just namespaced global variables and/or global functions).

What makes OO what is it - is actually dynamic dispatching, not that "everything is an object". A type is "a" object in Ruby, sure - but a type is not "the" object (aka. - you're replacing OO with C-like methods and structures).

> Having a thousand different error classes
Correct. That's not the point. The point is catching e.g. `Jekyll::ConversionError => e` and having `#file_name` and `#line_number` and `#plugin_name`, while `Jekyll:BuildError` wouldn't have a `#line_number` or `#plugin_name`, because those wouldn't make sense.

In other words - if I catch `Jekyll:ConversionError` I expect a file name file number (if possible - of it's only available in a subclass).

This is because Haml operates on stdin, and so it doesn't have a file number. Jekyll:Converter knows the plugin name and line number and file name - but Jekyll::Build doesn't have that information. 

Which many anyone writing a tool to "jump to error line, regardless of plugin/templating engine used" ... would be screwed without Jekyll::ConversionError instances specifically (and the plugin name could help select the right syntax highlighting).

If Jekyll handles e.g. 100 layers, it makes sense to have 100 different exception types - with specific info for each layer.

Having one exception isn't enough, because Jekyll doesn't do "one thing".

API consumers aside, there's a problem of reporting errors to users (users are the final "exception handler").

Typically, user errors should be presented in the most outer layer - e.g. the CLI. This means to give accurate hints to users, you have to carry intimate, context-specific info from the guts of the app to to an unrelated layer ...

... and nothing beats proper, structured exception handling for doing this.

Most "beginner" programmers are "terrified" of exceptions, so they never get deep into the subject to design things properly upfront ...

... and then you end re-architecture a whole project just to display a simple user message.

The result: no one wants to do the rework, development grinds to a hald - and people get frustrated and create their own frameworks. (Or maintainers like me end up pouring massive amounts of "unappreciated" hours across many projects - just to get a simple user message working).

I appreciate you engaging in this discussion - it means a lot to me, since there's either at least some truth to both sides or we both gain knowledge we didn't have before.

My experience is: the lack of structured exceptions in complex projects just causes unecessary grief and wasted lives on pointless problems causes by "why so many classes instead of one simple method X?" (I'm not picking on you hear - it's just too common to fall for the "let's avoid changing the API at all costs" mentality, especially with SemVer being so important.

**Comment by envygeeks** [comment]
Jekyll is not a complex project.

-

> A type is not the same as an object.

Actually, in Ruby types are objects... but that's not even relevant here.  I think you might have confused a `type` method that describes what the error type and asserts where an error happens by saying `Jekyll::Error#type => :convertor` `Jekyll::Error#original_exception => <Exception>` Jekyll::Error#line => 10` with primitive types and then turned what was a simple conversation into something entirely off topic and a debate that doesn't even matter here.  I don't even know where or what you are on about because you are fatally trying to drill down a complex definition to a simplistic fundamental feature, that is not the only primary concern of OOP and it's not even relevant here... at all.

- 

As I stated before I digress from this design decision and doubly so as I think there is a bit of confusion going on here but I would like to note that the need for simplicity has little (if ever to do) with the want to keep an API the same because of SemVer and the need to keep an API stable has little to do with SemVer more than it has to do with the fact that many people consume our API's in one way or another and major changes need happen with warning.

-

I am fine with either one, including a compromise to wrap Jekyll::Convertor::Error with Jekyll::Error and have Jekyll::Convertor::Error do all the work and Jekyll::Error raise a consistent error API across multiple error types.

**Comment by e2** [comment]
> Jekyll is not a complex project.

If you believe that, try improving the architecture and releasing an improved version without introducing bugs or breaking SemVer and without creating migration headaches for users and API consumers.

Sure, it's not Rails, but here's a list of classes:

Here's a list of classes:
```bash
$ ack 'class '
lib/jekyll/cleaner.rb
4:  class Site
6:    class Cleaner
lib/jekyll/collection.rb
2:  class Collection
lib/jekyll/command.rb
2:  class Command
4:    class << self
lib/jekyll/commands/build.rb
3:    class Build < Command
5:      class << self
71:      end # end of class << self
lib/jekyll/commands/docs.rb
3:    class Docs < Command
5:      class << self
lib/jekyll/commands/doctor.rb
3:    class Doctor < Command
4:      class << self
lib/jekyll/commands/help.rb
3:    class Help < Command
4:      class << self
lib/jekyll/commands/new.rb
5:    class New < Command
6:      class << self
lib/jekyll/commands/serve.rb
4:    class Serve < Command
6:      class << self
lib/jekyll/configuration.rb
4:  class Configuration < Hash
lib/jekyll/converter.rb
2:  class Converter < Plugin
lib/jekyll/converters/identity.rb
3:    class Identity < Converter
lib/jekyll/converters/markdown/kramdown_parser.rb
3:    class Markdown
4:      class KramdownParser
lib/jekyll/converters/markdown/maruku_parser.rb
3:    class Markdown
4:      class MarukuParser
lib/jekyll/converters/markdown/rdiscount_parser.rb
3:    class Markdown
4:      class RDiscountParser
lib/jekyll/converters/markdown/redcarpet_parser.rb
3:    class Markdown
4:      class RedcarpetParser
lib/jekyll/converters/markdown.rb
3:    class Markdown < Converter
71:      # Private: Determine whether a class name is an allowed custom markdown
72:      # class name
lib/jekyll/converters/textile.rb
3:    class Textile < Converter
lib/jekyll/document.rb
4:  class Document
lib/jekyll/draft.rb
3:  class Draft < Post
lib/jekyll/entry_filter.rb
2:  class EntryFilter
lib/jekyll/errors.rb
3:    class FatalException < RuntimeError
6:    class MissingDependencyException < FatalException
lib/jekyll/excerpt.rb
4:  class Excerpt
lib/jekyll/frontmatter_defaults.rb
2:  # This class handles custom defaults for YAML frontmatter settings.
7:  class FrontmatterDefaults
lib/jekyll/generator.rb
2:  class Generator < Plugin
lib/jekyll/layout.rb
2:  class Layout
lib/jekyll/layout_reader.rb
2:  class LayoutReader
lib/jekyll/log_adapter.rb
2:  class LogAdapter
lib/jekyll/page.rb
2:  class Page
lib/jekyll/plugin.rb
2:  class Plugin
9:    # Fetch all the subclasses of this class and its subclasses' subclasses.
52:    # other - The class to be compared.
61:    # other - The class to be compared.
65:      self.class <=> other.class
lib/jekyll/plugin_manager.rb
2:  class PluginManager
lib/jekyll/post.rb
2:  class Post
lib/jekyll/publisher.rb
2:  class Publisher
lib/jekyll/related_posts.rb
2:  class RelatedPosts
4:    class << self
lib/jekyll/renderer.rb
4:  class Renderer
lib/jekyll/site.rb
5:  class Site
403:    # Get the implementation class for the given Converter.
409:      matches = converters.select { |c| c.class == klass }
417:    # Create array of instances of the subclasses of the class or module
420:    # klass - class or module containing the subclasses which should be
lib/jekyll/static_file.rb
2:  class StaticFile
lib/jekyll/stevenson.rb
2:  class Stevenson < ::Logger
lib/jekyll/tags/highlight.rb
3:    class HighlightBlock < Liquid::Block
lib/jekyll/tags/include.rb
5:    class IncludeTagError < StandardError
14:    class IncludeTag < Liquid::Tag
154:    class IncludeRelativeTag < IncludeTag
lib/jekyll/tags/post_url.rb
3:    class PostComparer
44:    class PostUrl < Liquid::Tag
lib/jekyll/url.rb
13:  class URL
lib/jekyll.rb
71:  class << self
```

And by contrast here are the error classes:
```bash
$ ack 'class '|grep -i error
lib/jekyll/errors.rb:3:    class FatalException < RuntimeError
lib/jekyll/errors.rb:6:    class MissingDependencyException < FatalException
lib/jekyll/tags/include.rb:5:    class IncludeTagError < StandardError
```

> Jekyll::Error#type => :convertor

That's a bad practice no matter how you look at it. That's why I went into the discussion - this idea breaks too many established and expected paradigms.

Why add a `case` statement inside a `rescue` clause when Ruby already internally matches exceptions to the right clause anyway? 

It only makes exception handling more complex (from being type safe to being an artibrary `:symbol` with no upside).

I don't see how this is simpler in any way - it just creates a maintenance nightmare down the road.

More convenient *handling* of exceptions benefits Jekyll, because it could isolate and decouple it's own layers/classes. The point of having `original_error` is only as a transition strategy to avoid losing important information (until there's a better way of passing it).

**Comment by envygeeks** [comment]
This ticket has gone from productive to an attack, we understand your request and will work on completing it if we can, and have talked and will let other maintainers contribute to the discussion on what they think, I have stated my piece. Sorry for the accidental lock.

**Comment by e2** [comment]
@envygeeks - I apologize. The discussion turned complex and lengthy, I tried to be blunt and to the point technically.

**Comment by e2** [comment]
I mentioned the practice is bad (because it is, technically). It was meant to be technical critique, not a personal attack.

You are responsible for publicly presenting your opinions and professionalism. If you take other people's opinions about your solutions as a personal attack on you - I'm ending this discussion.

**Comment by envygeeks** [comment]
I never said it was a personal attack, I said it was an attack and I will leave the merits of design decisions to others, I will not argue solutions that where optional, not required and up for debate, and I will certainly no longer argue about Ruby's internals and OOP because of reasons I do not wish to state.  That said, I do like the idea and if I have time I'll work with @parkr on implementing it sometime when other features that we are working on aren't on the backlog.

**Comment by e2** [comment]
> I said it was an attack

If a contributor's/maintainer's/user's experience if Jekyll is negative, how I can I present that without it being seen as an "attack"?

I listed the classes to prove that Jekyll state is beyond "trivial", given how widespread it is - and how the exception handling is almost non-existent. What I wanted to show by that is not that it's "bad", but adding 3 Exception classes isn't really hurting simplicity in this case.

It just boggles my mind how adding 3 Exception classes is "complexity" - given the existing code by comparison (which is why I pasted the code).

If me trying to prepare a quality pull request for this (by discussing first, estimating the response, the need, the appreciation) is countered with solutions that are poor practices from my perspective ... I'm going to "give up" on investing the time to fix/test/debug Jekyll (I've had my share) and try other solutions instead.

I'd like to make Jekyll better (even to save myself lots of wasted hours in the future - due to having the same problems over and over again).

Don't take that as a "threat" that "if you don't treat me nicely, I won't contribute". 

It's more along the lines - "oh, why bother with fixes/testing/debugging/contributing/helping others with Jekyll issue -  since there's so much friction even when I suggest something I've put a lot of thought into".

You don't have time to implement this, and I don't want this to be an uphill experience for me. I honestly do want to improve the quality of Jekyll at this point (or I wouldn't bother with the discussion) - so this whole issue has me rolling my eyes convincing me I shouldn't bother.

I just wanted to communicate that, because I think it would be useful for you to know where I'm coming from.
